### PR TITLE
Check the code versioning before starting the upgrade process

### PR DESCRIPTION
**Pull request by christiancp100**
### What does it do?

It checks if the code is under a git versioning and if it is clean before continuing with the update process

### Why is it needed?

To make the upgrade safer

### How to test it?

- Run the upgrade in a process with a dirty git tree and see that there is a warning and asks you if you want to continue
- Run the upgrade in a project with no Git VCS and see there is a warning and asks you if you want to continue
- Run the upgrade with --force option and it should continue no matter the 2 previous scenarios

### CODE REVIEW

File: packages/utils/upgrade/package.json [hunk 1]
```diff
@@ -82,4 +83,4 @@
     "node": ">=18.0.0 <=20.x.x",
     "npm": ">=6.0.0"
   }
-}
+}
```
Comment (innerdvations) [inline packages/utils/upgrade/package.json:86]:
```suggestion
}

```

File: packages/utils/upgrade/src/core/requirements/is-clean-git-repo.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+import { simpleGit } from 'simple-git';
+import type { TaskOptions } from '../../types';
+
+type Params = Pick<TaskOptions, 'confirm' | 'logger'> & {
+  force: boolean;
+  cwd: string;
+};
+
+export const isCleanGitRepo = async ({ cwd, logger, force, confirm }: Params) => {
+  const git = simpleGit({ baseDir: cwd });
```
Comment (innerdvations) [inline packages/utils/upgrade/src/core/requirements/is-clean-git-repo.ts:11]: According to their docs, git is a pre-req. Could you add an extra check for `hasGitInstalled`?

**Review by innerdvations — APPROVED** [review]
LGTM!

File: packages/utils/upgrade/src/core/requirements/is-clean-git-repo.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    isRepo: true,
+    isClean: true,
+    isGitInstalled: true,
+  };
+
+  try {
+    // Check if Git is installed
+    await git.version();
+    // Check if the path is under version control
+    repoStatus.isRepo = await git.checkIsRepo();
```
Comment (Convly) [inline packages/utils/upgrade/src/core/requirements/is-clean-git-repo.ts:22]: Does this check if the current working directory contains a .git folder or does it detect .git folder higher in the file structure (e.g. mono-repositories?)

**Review by Convly — COMMENTED** [review]
Two things:
- I would confirm in between each requirement's fails & ask before continuing
- It could make sense to create an abstraction to check any kind of requirements (we'll need to add `node`, etc... in the future). But that can be part of the next PR
- For some reason it feels weird for the git requirement to be given the `confirm` function as a parameter. I would rather try to find a solution to keep it in the task itself

All of the above can fit in a new iteration, this is a good first step. The simpleGit API looks clean.

File: packages/utils/upgrade/src/core/requirements/is-clean-git-repo.ts (hunk 3 above)
Comment (christiancp100) [reply packages/utils/upgrade/src/core/requirements/is-clean-git-repo.ts:22]: It works with the mono repo too

**Review by Convly — CHANGES_REQUESTED** [review]
It seems that the added unit tests are failing

**Comment by innerdvations** [comment]
> * For some reason it feels weird for the git requirement to be given the `confirm` function as a parameter. I would rather try to find a solution to keep it in the task itself

Similarly, the force option should be passed from the CLI, but that can also be a follow-up PR.

File: packages/utils/upgrade/src/__tests__/core/requirements/is-clean-git-repo.test.ts [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    expect(mockSimpleGit.checkIsRepo).toHaveBeenCalled();
+    expect(mockSimpleGit.status).not.toHaveBeenCalled();
+    expect(console.warn).toHaveBeenCalledWith(
+      chalk.yellow(`[WARN]\t[${isoString}]`),
+      'Unable to proceed with the upgrade:'
+    );
+
+    expect(console.warn).toHaveBeenCalledWith(
+      chalk.yellow(`[WARN]\t[${isoString}]`),
+      '- The codebase is not under version control.'
```
Comment (innerdvations) [inline packages/utils/upgrade/src/__tests__/core/requirements/is-clean-git-repo.test.ts:131]:
```suggestion
      '- No git repository was detected in the directory.'
```

File: packages/utils/upgrade/src/__tests__/core/requirements/is-clean-git-repo.test.ts [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
```
Comment (innerdvations) [inline packages/utils/upgrade/src/__tests__/core/requirements/is-clean-git-repo.test.ts:111]:
```suggestion
      '- No git repository was detected in the directory.'
```

**Comment by innerdvations** [comment]
Updated test text, should pass now.

**Comment by innerdvations** [comment]
@Convly all good to merge or do you want any of those 3 points addressed in this PR?

**Comment by christiancp100** [comment]
> @Convly all good to merge or do you want any of those 3 points addressed in this PR?

I've talked to him and I'll try to address the confirm one, but the one related to the confirmation on each step is up to discussion on monday

**Comment by Convly** [comment]
We can merge as is, let's implement the rest of the modifications later

**Pull request merged**